man/anno.man
.TH ANNO %manext1% 2005-12-07 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
anno \- annotate nmh messages
.SH SYNOPSIS
.HP 5
.na
.B anno
.RB [ \-help ]
.RB [ \-version ]
.RI [ +folder ]
.RI [ msgs ]
.RB [ \-component
.IR field ]
.RB [ \-inplace " | " \-noinplace ]
.RB [ \-date " | " \-nodate ]
.RB [ \-draft ]
.RB [ \-append ]
.RB [ \-list ]
.RB [ \-delete ]
.RB [ \-number
.IR [ num|all ]]
.RB [ \-preserve " | " \-nopreserve ]
.RB [ \-text
.IR body ]
.ad
.SH DESCRIPTION
.B anno
manipulates header fields or
.I annotations
in messages.
Header fields consist of a field name and an optional field body
as defined by RFC 2822.
The
.B -component
switch specifies the field name, and the
.B -text
switch specifies the field body.
.PP
The messages are either the
.I msgs
in the named folder, or the draft if invoked with the
.B -draft
switch.
.PP
Usually, annotation is performed by the commands
.BR dist ,
.BR forw ,
and
.BR repl ,
if they are given the
.B \-anno
switch.  This allows you to keep track of your distribution of,
forwarding of, and replies to a message.
.PP
By using
.BR anno ,
you can perform arbitrary annotations of your own.
Each message selected will be annotated with the lines
.PP
    field:\ date
    field:\ body
.PP
The
.B \-nodate
switch inhibits the date annotation, leaving only the
body annotation.
.PP
By default,
.B anno
prepends the annotations to the message.
Annotations are instead appended if the
.B -append
switch is specified.
.PP
If a
.B \-component
.I field
is not specified when
.B anno
is invoked,
.B anno
will prompt the user for a field name.
.PP
The field specified must be a valid RFC 2822-style message field name,
consisting only of alphanumeric characters and dashes.
The body may consist of arbitrary text.
.PP
Normally
.B anno
does the annotation in place in order to preserve
any links to the message.  You may change this by using the
.B \-noinplace
switch.
.PP
The
.B -list
switch produces a listing of the field bodies for header fields with
names matching the specified component, one per line.
The listing is numbered, starting at 1, if the
.B -number
switch is also used.
A tab character separates the number and the field body.
The field body is treated as if it is a file name, and only the final
path name component is listed.
The complete field body is listed if the
.B -text
switch is used; the argument to the
.B -text
switch is required but ignored.
.PP
The
.B -delete
switch removes header fields from messages.
The first header field whose name matches the component is deleted if
no other options are specified.
If the
.B -text
switch is used in conjunction with the
.B -delete
switch, the first header field whose name matches the component and
whose body matches the text is deleted.
The text is treated as if it was a file name; if it begins with a
slash, the entire field body must match the text, otherwise just the
last path name component of the field body must match.
If the
.B -number
switch is used in conjunction with the
.B -delete
switch, header field
.I num
whose name matches the component is deleted.
The number matches that which is produced by the
.B -list
switch.
The special value
.B all
can be used for the number, and causes all components that match the
name to be deleted.
.PP
By default,
.B anno
changes the last-accessed and last-modified times on annotated messages
to the time of the annotation.
.B anno
preserves the original times if the
.B -preserve
switch is used.
A matching
.B -nopreserve
switch allows time preservation to be turned off, if enabled
in the profile.
.SH FILES
.TP 20
$HOME/.mh_profile
The user's profile.
.SH "PROFILE COMPONENTS"
.PD 0
.TP 20
Path:
To determine the user's nmh directory.
.TP
Current\-Folder:
To find the default current folder.
.PD
.SH "SEE ALSO"
.IR dist (1),
.IR forw (1),
.IR repl (1)
.SH DEFAULTS
.PD 0
.TP 20
+folder
The current folder.
.TP
msgs
The current message.
.TP
-inplace
.TP
-date
.PD
.SH CONTEXT
If a folder is given, it will become the current folder.  The first
message annotated will become the current message.


NEWS
#
# NEWS - Release notes for nmh 1.6
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been nearly two years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

------------
NEW FEATURES
------------

The following are new features for the 1.6 release of nmh:

- All nmh programs will now warn about multiple profile entries for
  the same non-null, non-comment component.  In previous versions, all
  but the first were silently ignored.
- mhmail now supports all post(8) options, or optionally can use
  send(1) with all of its options.  Its formerly undocumented -resent
  switch has been documented.  And it has new -attach, -headerfield,
  and -send/-nosend switches.
- Default to flock() locking on OpenBSD and Darwin.
- Added -directives support to mhbuild(1) to control whether or not
  mhbuild will honor MIME directives by default.  And added support
  for special #on/#off/#pop directives to control the MIME directive
  processing state.
- Added -messageid switch to send(1) and post(8).  This allows
  selection of the style to use for generated Message-ID and
  Content-ID header fields.  The default localname style is
  pid.time@localname, where time is in seconds, and matches previous
  behavior.  The random style replaces the localname with some
  (pseudo)random bytes and uses microsecond-resolution time.
- Added -clobber switch to mhstore(1) to control overwriting of
  existing files.
- Added -outfile switch to mhstore(1).
- Added -noall/-all switches to sortm(1).  sortm -noall requires
  a messages argument.
- $PAGER overrides the compiled-in default pager command.
- Added etc/scan.highlighted format file, as an example of how to
  highlight/colorize the output of scan(1).
- inc(1) now supports a -port switch to specify the port used by the
  POP server.
- pick(1) now decodes MIME-encoded header fields before searching.
- The VISUAL and EDITOR environment variables are now supported as fallbacks
  if the user does not configure an editor entry in their profile.
- The format engine (mh_format(5)) now properly accounts for multibyte
  characters when accounting for column widths.
- burst(1) now can burst MIME-formatted digests (messages that contain
  message/rfc822 parts instead of messages formatted with RFC 934).
- All proc entries (showproc, moreproc, etc) can now accept entries that
  contain spaces and shell metacharacters.  If found, such entries will
  either be space-split or processed by /bin/sh.
- A new program, fmttest(1) is included to help debug format files.  See
  NOTABLE CHANGES below.
- mhshow/mhstore now have support for RFC-2017 (access-type=url) for
  external message bodies.
- Added -retainsequences switch to refile(1).
- A new program, mhfixmsg(1), is included to rewrite MIME messages with
  various transformations.  See NOTABLE CHANGES below.
- Added -[no]rmmproc switches to rmm(1).
- Added support for Content-Disposition header (RFC 2183) to mhstore(1)
  and mhn(1) when used with -auto.
- All nmh commands now support transactional locking for sequence files.
  See NOTABLE CHANGES below.
- There is no longer a per-folder maximum number of sequences.
- For the SMTP MTA TLS can now be negotiated at the beginning of the
  connection with the -initialtls switch.
- Messages can now be selected using a relative offset from some other
  message, or from the start or end of a sequence.  See mh-sequence(5).
- The -changecur and -nochangecur switches have been added to mhlist(1).
- mhbuild(1) can now encode 8-bit message headers using RFC-2047 encoding
  rules.  See NOTABLE CHANGES below.
- The whatnow(1) attach feature will determine the content type of an
  attachment using a program such as file --mime-type, if available at
  configuration time.  If not, it will use mhshow-suffix- entries as
  before.  The -v switch to attach causes it to display the mhbuild
  directive that send(1) will use.
- mhbuild(1) now supports the -auto/-noauto flags (to be used by send(1)
  when invoking mhbuild automatically).
- mhbuild(1) now is automatically run by send, to insure that all outgoing
  messages have proper MIME formatting.  See NOTABLE CHANGES below.
- A new header, "Attach", is supported by mhbuild; it is used to replace
  previous functionality (which by default used a header named
  Nmh-Attachment).  See NOTABLE CHANGES below.
- The default Content-Transfer-Encoding for text parts is now 8bit.
- mhbuild(1) now supports a selectable Content-Transfer-Encoding
- If nmh was configured with iconv(3) support, mhshow will convert, if
  necessary, the charset of text/plain content to match the user's
  locale setting.  See NOTABLE CHANGES below.
- Added support for %{charset} display string escape to mhshow(1).
- The MIME parsing and generating routines now support RFC 2231 extended
  parameter information.
- mh-mime(7) now provides an introduction to nmh's MIME handling.
- mhshow(1) will now by default display all text content under one pager,
  and display markers for non-text and non-inline content.  The content
  markers are changeable via mh-format(5).  See NOTABLE CHANGES below.

-----------------
OBSOLETE FEATURES
-----------------
- Changed exit status of each nmh command's -version and -help
  switches from 1 to 0.
- The following environment variables were deprecated in nmh 1.5
  and removed from this release:
    MHPOPDEBUG (use -snoop command line switch instead)
    MM_NOASK (use -nolist and -nopause command line switches instead)
    NOMHNPROC (use -nocheckmime command line switch instead)
    FACEPROC (undocumented faceproc feature removed)
- Changed repl and dist default to -noatfile.  The default of -atfile

uip/anno.c
/* anno.c -- annotate messages
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 *
 *	Four new options have been added: delete, list, number, and draft.
 *	Message header fields are used by the new MIME attachment code in
 *	the send command.  Adding features to generalize the anno command
 *	seemed to be a better approach than the creation of a new command
 *	whose features would overlap with those of the anno command.
 *
 *	The -draft option causes anno to operate on the current draft file
 *	instead of on a message sequence.
 *
 *	The -delete option deletes header elements that match the -component
 *	field name.  If -delete is used without the -text option, the first
 *	header field whose field name matches the component name is deleted.
 *	If the -delete is used with the -text option, and the -text argument
 *	begins with a /, the first header field whose field name matches the
 *	component name and whose field body matches the text is deleted.  If
 *	the -text argument does not begin with a /, then the text is assumed
 *	to be the last component of a path name, and the first header field
 *	whose field name matches the component name and a field body whose
 *	last path name component matches the text is deleted.  If the -delete
 *	option is used with the new -number option described below, the nth
 *	header field whose field name matches the component name is deleted.
 *	No header fields are deleted if none of the above conditions are met.
 *
 *	The -list option outputs the field bodies from each header field whose
 *	field name matches the component name, one per line.  If no -text
 *	option is specified, only the last path name component of each field
 *	body is output.  The entire field body is output if the -text option
 *	is used; the contents of the -text argument are ignored.  If the -list
 *	option is used in conjunction with the new -number option described
 *	below, each line is numbered starting with 1.  A tab separates the
 *	number from the field body.
 *
 *	The -number option works with both the -delete and -list options as
 *	described above.  The -number option takes an optional argument.  A
 *	value of 1 is assumed if this argument is absent.
 */

#include <h/mh.h>
#include <h/utils.h>

#define ANNO_SWITCHES \
    X("component field", 0, COMPSW) \
    X("inplace", 0, INPLSW) \
    X("noinplace", 0, NINPLSW) \
    X("date", 0, DATESW) \
    X("nodate", 0, NDATESW) \
    X("text body", 0, TEXTSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \
    X("draft", 2, DRFTSW) \
    X("list", 1, LISTSW) \
    X("delete", 2, DELETESW) \
    X("number", 2, NUMBERSW) \
    X("append", 1, APPENDSW) \
    X("preserve", 1, PRESERVESW) \
    X("nopreserve", 3, NOPRESERVESW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(ANNO);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(ANNO, switches);
#undef X

/*
 * static prototypes
 */
static void make_comp (char **);


int
main (int argc, char **argv)
{
    int inplace = 1, datesw = 1;
    int msgnum;
    char *cp, *maildir;
    char *comp = NULL, *text = NULL, *folder = NULL, buf[BUFSIZ];
    char **argp, **arguments;
    struct msgs_array msgs = { 0, 0, NULL };
    struct msgs *mp;
    int		append = 0;		/* append annotations instead of default prepend */
    int		delete = -2;		/* delete header element if set */
    char	*draft = NULL;	/* draft file name */
    int		isdf = 0;		/* return needed for m_draft() */
    int		list = 0;		/* list header elements if set */
    int		number = 0;		/* delete specific number of like elements if set */

    if (nmh_init(argv[0], 1)) { return 1; }

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW:
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW:
		    adios (NULL, "-%s unknown", cp);

		case HELPSW:
		    snprintf (buf, sizeof(buf), "%s [+folder] [msgs] [switches]",
			invo_name);
		    print_help (buf, switches, 1);
		    done (0);
		case VERSIONSW:
		    print_version(invo_name);
		    done (0);

		case COMPSW:
		    if (comp)
			adios (NULL, "only one component at a time!");
		    if (!(comp = *argp++) || *comp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    continue;

		case DATESW:
		    datesw++;
		    continue;
		case NDATESW:
		    datesw = 0;
		    continue;

		case INPLSW:
		    inplace++;
		    continue;
		case NINPLSW:
		    inplace = 0;
		    continue;

		case TEXTSW:
		    if (text)
			adios (NULL, "only one body at a time!");
		    if (!(text = *argp++) || *text == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    continue;

		case DELETESW:		/* delete annotations */
		    delete = 0;
		    continue;

		case DRFTSW:		/* draft message specified */
		    draft = "";
		    continue;

		case LISTSW:		/* produce a listing */
		    list = 1;
		    continue;

		case NUMBERSW:		/* number listing or delete by number */
		    if (number != 0)
			adios (NULL, "only one number at a time!");

		    if (argp - arguments == argc - 1 || **argp == '-')
			number = 1;
		    
		    else {
		    	if (strcmp(*argp, "all") == 0)
		    	    number = -1;

		    	else if (!(number = atoi(*argp)))
			    adios (NULL, "missing argument to %s", argp[-1]);

			argp++;
		    }

		    delete = number;
		    continue;

		case APPENDSW:		/* append annotations instead of default prepend */
		    append = 1;
		    continue;

		case PRESERVESW:	/* preserve access and modification times on annotated message */
		    annopreserve(1);
		    continue;

		case NOPRESERVESW:	/* don't preserve access and modification times on annotated message (default) */
		    annopreserve(0);
		    continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		adios (NULL, "only one folder at a time!");
	    else
		folder = pluspath (cp);
	} else
		app_msgarg(&msgs, cp);
    }

    /*
     *	We're dealing with the draft message instead of message numbers.
     *	Get the name of the draft and deal with it just as we do with
     *	message numbers below.
     */

    if (draft != NULL) {
	if (msgs.size != 0)
	    adios(NULL, "can only have message numbers or -draft.");

	draft = getcpy(m_draft(folder, NULL, 1, &isdf));

	make_comp(&comp);

	if (list)
	    annolist(draft, comp, text, number);
	else
	    annotate (draft, comp, text, inplace, datesw, delete, append);

	done(0);
	return 1;
    }

    if (!context_find ("path"))
	free (path ("./", TFOLDER));
    if (!msgs.size)
	app_msgarg(&msgs, "cur");
    if (!folder)
	folder = getfolder (1);
    maildir = m_maildir (folder);

    if (chdir (maildir) == NOTOK)
	adios (maildir, "unable to change directory to");

    /* read folder and create message structure */
    if (!(mp = folder_read (folder, 1)))
	adios (NULL, "unable to read folder %s", folder);

    /* check for empty folder */
    if (mp->nummsg == 0)
	adios (NULL, "no messages in %s", folder);

    /* parse all the message ranges/sequences and set SELECTED */
    for (msgnum = 0; msgnum < msgs.size; msgnum++)
	if (!m_convert (mp, msgs.msgs[msgnum]))
	    done (1);

    make_comp (&comp);

    /* annotate all the SELECTED messages */
    for (msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++) {
	if (is_selected(mp, msgnum)) {
	    if (list)
		annolist(m_name(msgnum), comp, text, number);
	    else
		annotate (m_name (msgnum), comp, text, inplace, datesw, delete, append);
	}
    }

    context_replace (pfolder, folder);	/* update current folder  */
    seq_setcur (mp, mp->lowsel);	/* update current message */
    seq_save (mp);	/* synchronize message sequences */
    folder_free (mp);	/* free folder/message structure */
    context_save ();	/* save the context file         */
    done (0);
    return 1;
}

static void
make_comp (char **ap)
{
    char *cp, buffer[BUFSIZ];

    if (*ap == NULL) {
	printf ("Enter component name: ");
	fflush (stdout);

	if (fgets (buffer, sizeof buffer, stdin) == NULL)
	    done (1);
	*ap = trimcpy (buffer);
    }

    if ((cp = *ap + strlen (*ap) - 1) > *ap && *cp == ':')
	*cp = 0;
    if (!**ap)
	adios (NULL, "null component name");
    if (**ap == '-')
	adios (NULL, "invalid component name %s", *ap);
    if (strlen (*ap) >= NAMESZ)
	adios (NULL, "too large component name %s", *ap);

    for (cp = *ap; *cp; cp++)
	if (!isalnum ((unsigned char) *cp) && *cp != '-')
	    adios (NULL, "invalid component name %s", *ap);
}
